### PR TITLE
feat(personal-representative): Service API extended to include personal reps and rights

### PR DESCRIPTION
**Pull request by FridrikMProgramm**
# Personal Representative - ServiceAPI Extended to include personal reps and rights
https://programm.jira.com/browse/SSI-6

## What
Personal Representatvie Service API.
Extended to allow maintaining personal representatives and their rights.

## Why
Extended to allow ministry of welfare to add and remove personal representative and adjust their rights.

## Screenshots
image

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: apps/services/personal-representative/src/app/modules/personalRepresentative/personalRepresentative.controller.ts [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+@ApiBearerAuth()
+export class PersonalRepresentativeController {
+  constructor(
+    @Inject(PersonalRepresentativeService)
+    private readonly prService: PersonalRepresentativeService,
+    private readonly auditService: AuditService,
+  ) {}
+
+  /** Gets all persoanal representatives */
+  @ApiOperation({ summary: 'Gets all persoanal representatives' })
```
Comment (fjandakornid) [inline apps/services/personal-representative/src/app/modules/personalRepresentative/personalRepresentative.controller.ts:52]: Typo: persoanal

File: libs/auth-api-lib/src/lib/entities/models/personal-representative.model.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import { PersonalRepresentativeRight } from './personal-representative-right.model'
+import { PersonalRepresentativeDTO } from '../dto/personal-representative.dto'
+
+@Table({
+  tableName: 'personal_representative',
+})
+export class PersonalRepresentative extends Model<PersonalRepresentative> {
+  @PrimaryKey
+  @Column({
+    type: DataType.STRING,
```
Comment (fjandakornid) [inline libs/auth-api-lib/src/lib/entities/models/personal-representative.model.ts:21]: Could let the database generate UUIDs:

@Column({
    type: DataType.UUID,
    primaryKey: true,
    allowNull: false,
    defaultValue: DataType.UUIDV4,
  })
  @ApiProperty()
  id!: string

**Review by fjandakornid — COMMENTED** [review]
Few comments

File: libs/auth-api-lib/src/lib/entities/models/personal-representative.model.ts (hunk 2 above)
Comment (FridrikMProgramm) [reply libs/auth-api-lib/src/lib/entities/models/personal-representative.model.ts:21]: Will explore in next PR.

**Review by eirikurn — CHANGES_REQUESTED** [review]
When you get to the API, please remember to write API-based unit tests.

File: libs/auth-api-lib/src/lib/entities/models/personal-representative-right.model.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  @ForeignKey(() => PersonalRepresentativeRightType)
+  @Column({
+    type: DataType.STRING,
+    allowNull: false,
+  })
+  @ApiProperty()
+  rightTypeCode!: string
+
+  @ApiProperty({
+    type: () => [PersonalRepresentative],
```
[the remark this answers is not in the document]
Comment (FridrikMProgramm) [reply libs/auth-api-lib/src/lib/entities/models/personal-representative-right.model.ts:54]: No this is the connection table between PersonalRep and PersonalRepRightType. Each record connects to gogether one of each.

File: libs/auth-api-lib/src/lib/entities/models/personal-representative-right.model.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    required: true,
+  })
+  @BelongsTo(() => PersonalRepresentative)
+  personalRepresentative?: unknown
+
+  @ApiProperty({
+    type: () => [PersonalRepresentativeRightType],
```
[the remark this answers is not in the document]
Comment (FridrikMProgramm) [reply libs/auth-api-lib/src/lib/entities/models/personal-representative-right.model.ts:61]: Same here

File: libs/auth-api-lib/src/lib/services/personalRepresentative.service.ts [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+          },
+        ],
+        where: {
+          [Op.or]: [
+            {
+              name: { [Op.like]: searchString },
+            },
+            {
+              code: { [Op.like]: searchString },
+            },
```
[the remark this answers is not in the document]
Comment (FridrikMProgramm) [reply libs/auth-api-lib/src/lib/services/personalRepresentative.service.ts:197]: Should have been deeper in the include tree.

File: libs/auth-api-lib/src/lib/entities/models/personal-representative-right.model.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (eirikurn) [reply libs/auth-api-lib/src/lib/entities/models/personal-representative-right.model.ts:54]: Yes I mean you are returning `[PersonalRepresentative]` instead of just `PersonalRepresentative`, which I think is reflected as an array of PersonalRepresentatives in the OpenAPI schema.

File: apps/services/personal-representative/src/app/modules/personalRepresentative/e2e/personalRepresentative.spec.ts [hunk 6]
```diff
@@ -170,11 +171,11 @@ describe('PersonalRepresentativeController', () => {
 
       // Test get personal rep
       const response = await server
-        .get(`/v1/personal-representative/all/`)
+        .get(`/v1/personal-representative/all/false?limit=1`)
```
Comment (eirikurn) [inline apps/services/personal-representative/src/app/modules/personalRepresentative/e2e/personalRepresentative.spec.ts:174]: Ah, just noticed this URL structure now.

We should follow our API Design Guide when it comes to URL [structure](https://docs.devland.is/technical-overview/api-design-guide/resource-oriented-design#resources) and [naming](https://docs.devland.is/technical-overview/api-design-guide/naming-conventions#uris), especially since this is a Island.is API which we're sharing with other parties (difficult to "fix" later).

You can check out auth-public-api for inspiration, which is modeling similar things and was recently refactored to follow the design guide. You can also chat with @saevarma for input on specific challenges in the API design.

Comment (eirikurn) [reply apps/services/personal-representative/src/app/modules/personalRepresentative/e2e/personalRepresentative.spec.ts:174]: E.g. we should be able to use optional or required query parameters to replace some of these URL segments.

**Review by eirikurn — CHANGES_REQUESTED** [review]

File: libs/auth-api-lib/src/lib/entities/models/personal-representative-right.model.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (FridrikMProgramm) [reply libs/auth-api-lib/src/lib/entities/models/personal-representative-right.model.ts:54]: :/ my bad

File: libs/auth-api-lib/src/lib/services/personalRepresentative.service.ts [hunk 7]
```diff
@@ -21,49 +23,76 @@ export class PersonalRepresentativeService {
...
+      Model: this.personalRepresentativeModel,
+      limit: query.limit || 10,
+      after: query.after,
+      before: query.before,
+      primaryKeyField: 'id',
+      orderOption: [['id', 'DESC']],
+      where: whereClause,
+    })
+
+    // Since paginate in sequlize does not support inclue correctly we need to fetch the rights array separately
```
[the remark this answers is not in the document]
Comment (FridrikMProgramm) [reply libs/auth-api-lib/src/lib/services/personalRepresentative.service.ts]: I'll look into it when I get the chance. Will leave it like this for now.

File: apps/services/personal-representative/src/app/modules/rightTypes/rightTypes.controller.ts [hunk 8]
```diff
@@ -103,7 +103,7 @@ export class RightTypesController {
   @ApiOkResponse()
   async removeAsync(
     @Param('code') code: string,
-    @CurrentUser() user: User,
+    @CurrentActor() user: User,
```
Comment (eirikurn) [inline apps/services/personal-representative/src/app/modules/rightTypes/rightTypes.controller.ts:106]: Why this change? This should only be used in very special circumstances (APIs returning data for the real user even during delegation).

File: apps/services/personal-representative/src/app/modules/rightTypes/rightTypes.controller.ts [hunk 9]
```diff
@@ -131,7 +131,7 @@ export class RightTypesController {
   })
   async create(
     @Body() rightType: PersonalRepresentativeRightTypeDTO,
-    @CurrentUser() user: User,
+    @CurrentActor() user: User,
```
Comment (eirikurn) [inline apps/services/personal-representative/src/app/modules/rightTypes/rightTypes.controller.ts:134]: Same here

File: apps/services/personal-representative/src/app/modules/rightTypes/rightTypes.controller.ts [hunk 10]
```diff
@@ -158,7 +158,7 @@ export class RightTypesController {
   async update(
     @Param('code') code: string,
     @Body() rightType: PersonalRepresentativeRightTypeDTO,
-    @CurrentUser() user: User,
+    @CurrentActor() user: User,
```
Comment (eirikurn) [inline apps/services/personal-representative/src/app/modules/rightTypes/rightTypes.controller.ts:161]: Same here

File: libs/auth-api-lib/src/lib/services/personalRepresentative.service.ts [hunk 11]
```diff
@@ -49,14 +45,14 @@ export class PersonalRepresentativeService {
     const result = await paginate({
       Model: this.personalRepresentativeModel,
       limit: query.limit || 10,
-      after: query.after,
-      before: query.before,
+      after: query.after ?? '',
+      before: query.before ?? '',
```
Comment (eirikurn) [inline libs/auth-api-lib/src/lib/services/personalRepresentative.service.ts:49]: Interesting. This is mildly annoying.

File: libs/auth-api-lib/src/lib/services/personalRepresentativeRightType.service.ts [hunk 12]
```diff
@@ -20,13 +20,13 @@ export class PersonalRepresentativeRightTypeService {
   /** Get's all personal repreasentative right types  */
   async getMany(
     // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    query: any,
+    query: PaginationDto,
```
Comment (eirikurn) [inline libs/auth-api-lib/src/lib/services/personalRepresentativeRightType.service.ts:23]: Can remove the comment above :D

**Review by eirikurn — APPROVED** [review]
Wow, super quick turnaround with the URLs. Great job!

Just revert to CurrentUser and we're golden. 👍

File: apps/services/personal-representative/src/app/modules/rightTypes/rightTypes.controller.ts (hunk 8 above)
Comment (FridrikMProgramm) [reply apps/services/personal-representative/src/app/modules/rightTypes/rightTypes.controller.ts:106]: Was copying the auth-public-api but will change back

File: libs/auth-api-lib/src/lib/services/personalRepresentative.service.ts (hunk 11 above)
Comment (FridrikMProgramm) [reply libs/auth-api-lib/src/lib/services/personalRepresentative.service.ts:49]: Yes there is a mismatch in the PaginationDto class and the PaginateInput interface so this is needed unless we change the interface

**Review by eirikurn — APPROVED** [review]

**Review by fjandakornid — APPROVED** [review]
LGTM
